### ISSUE TITLE
Consider moving state to `$XDG_STATE_HOME` as opposed to `$XDG_CACHE_HOME`

### ISSUE THREAD

**Issue by shymega** [pimalaya/neverest#5]
To follow-up from our Matrix chat:

Currently, `neverest` stores state/locks in two places:
`$XDG_CACHE_HOME/pimalaya` and `$XDG_RUNTIME_DIR/pimalaya`.

The former is the one I'd like to see changed to `$XDG_STATE_HOME/pimalaya`, as
storing in `$XDG_CACHE_HOME/pimalaya` runs the risk of accidental deletion of
the state files.

The latter isn't as much of an issue.

We could, maybe in the future, store state in the IMAP/Maildir backends as well.

**Comment by codethief** [comment]
In how far is it safe vs. not safe to delete `$XDG_CACHE_HOME/pimalaya`? Could anything go wrong next time I run `neverest synchronize`? If yes, that folder should definitely be considered state.

**Comment by soywod** [comment]
> In how far is it safe vs. not safe to delete `$XDG_CACHE_HOME/pimalaya`? Could anything go wrong next time I run `neverest synchronize`? If yes, that folder should definitely be considered state.

The scariest scenario is when you delete your data, the `root-dir`, *without* deleting the cache. It may consider that all emails were removed and needs to be removed remotely.

If you delete the cache, Neverest will not be able to know if sth has been deleted right side or added left side. For now it comes with safe defaults: it will always take the path that does not delete data. Worst case you just get back emails you previously deleted, or get duplicates (same email in trash and inbox).

I believe cache is not crucial. It just helps the sync to take the right decision. Hence my initial choice behind the cache directory. That said, the state directory also makes sense, and may be a safer default. So let's do it!

**Comment by codethief** [comment]
@soywod Thanks for elaborating! That is quite scary indeed.

Would it make sense to add a "readonly" flag to protect either left or right side from accidental changes or a flag to indicate the "direction" of the sync? In my case I'm only ever interested in syncing from left to right.

**Comment by soywod** [comment]
> Would it make sense to add a "readonly" flag to protect either left or right side from accidental changes or a flag to indicate the "direction" of the sync?

You have permissions for that purpose: https://github.com/pimalaya/neverest/blob/92fa58db23cd6933c50e70f08318b042e87e1828/config.sample.toml#L82-L91

**Comment by shymega** [comment]
> In how far is it safe vs. not safe to delete `$XDG_CACHE_HOME/pimalaya`? Could anything go wrong next time I run `neverest synchronize`? If yes, that folder should definitely be considered state.

It's not safe to delete it IMO, given it stores state. It should be in the XDG state directory - cache is for temporary files, and admittedly - sometimes state - but for this use-case of neverest, this should not be considered safe.

**Comment by soywod** [comment]
Then I will fix it for the next release. Thank you for your feedback!

**Comment by shymega** [comment]
Did you mean to add the Invalid label? 🤔

**Comment by soywod** [comment]
Yes, in a way that "this implementation does not seem right". But not sure if it's a common practice, I can just remove it.

**Comment by soywod** [comment]
The `v1.0.0-rc` is available on `master`, using state instead of cache. You can download a pre-built binary [from the CI](https://github.com/pimalaya/neverest/actions/workflows/releases.yml): select the first green workflow and look for the *Artifacts* section. See also the [migration guide](https://github.com/pimalaya/neverest/blob/master/MIGRATION.md), for the config breaking changes. I will release as soon as possible.
